test/Index/index_objc_dynamic_refs.swift

```diff
@@ -0,0 +1,26 @@
+// RUN: %target-swift-ide-test -print-indexed-symbols -source-filename %s | %FileCheck %s
+// REQUIRES: objc_interop
+
+import Foundation
+
+@objc protocol AProtocol {
+    @objc optional func dynamicMethod()
+    // CHECK: [[@LINE-1]]:25 | instance-method/Swift | dynamicMethod() | [[DynamicMethod_USR:.*]] | Def
+    @objc optional var property: String { get }
+    // CHECK: [[@LINE-1]]:24 | instance-property/Swift | property | [[DynamicProperty_USR:.*]] | Def
+}
+
+class AClass {
+
+    weak var objcDelegate: AProtocol?
+    // CHECK: [[@LINE-1]]:14 | instance-property/Swift | objcDelegate | [[Delegate_USR:.*]] | Def
+
+    func doSomething() {
+        objcDelegate?.dynamicMethod?()
+        // CHECK: [[@LINE-1]]:23 | instance-method/Swift | dynamicMethod() | [[DynamicMethod_USR]] | Ref
+        // CHECK: [[@LINE-2]]:9 | instance-property/Swift | objcDelegate | [[Delegate_USR]] | Ref
```

Comment:
Notice that the order is wrong (the reference is coming before the base expression). I'm not sure why that's the case...

Reply:
It's because the base expression (`objcDelegate` in this case) is a child of the DynamicMemberRefExpr (`objcDelegate?.dynamicMethod`) in the AST and the new passReference() call for DynamicMemberRefExprs is happening in walkToExprPre, which the walker calls before visiting the expression's children.

To get it in source order you can do something similar to what's done in the handling of `MemberRefExpr` a bit above your change where it manually walks into the base expression before passing the reference, and then returns a value indicating that the walker shouldn't visit the expression's children:
```
    {
      ...

      // Visit in source order.
      if (!MRE->getBase()->walk(*this))
        return stopTraversal;
    }

    if (!passReference(MRE->getMember().getDecl(), MRE->getType(),
                       MRE->getNameLoc(),
                       ReferenceMetaData(SemaReferenceKind::DeclMemberRef,
                                         OpAccess)))
      return stopTraversal;

    // We already visited the children.
    return doSkipChildren();
```

Reply:
That worked, thanks! Pushed the improvement.